File: src/main/typescript/leetcode/_3_M.ts
function lengthOfLongestSubstring(s: string) {
    let res: number = 0;
    let existedIndexes: any = {};
    for (let left = -1, right = 0, sLen = s.length; right < sLen; right++) {
        const ei = existedIndexes[s.charAt(right)];
        left = Math.max(left, ei == null ? -1 : ei);
        existedIndexes[s.charAt(right)] = right;
        res = Math.max(res, right - left);
    }
    return res;
}

console.log(lengthOfLongestSubstring("test"));

File: python/leetcode/_3_M.py
from typing import Dict


class Solution(object):
    def lengthOfLongestSubstring(self, s: str):
        res = 0
        existedIndexes: Dict[str, int] = {}
        left = -1
        right = 0
        sLen = len(s)
        while right < sLen:
            ei = existedIndexes.get(s[right])
            if ei is None:
                ei = -1
            left = max(left, ei)
            existedIndexes[s[right]] = right
            res = max(res, right - left)
            right += 1
        return res

if __name__ == '__main__':
    print(Solution().lengthOfLongestSubstring("test"))
